### PR TITLE
Update documentation for workspace inheritance

### PR DESCRIPTION
**Pull request by Muscraft**
Tracking issue: #8415
RFC: rust-lang/rfcs#2906

This updates documentation about workspace inheritance in the Cargo Book. This is meant to move the documentation into a state that is acceptable to move after stabilization. It currently proposes adding sections to `workspaces.md` and `specifying-dependencies.md`.

r? @epage

### CODE REVIEW

File: src/doc/src/reference/unstable.md [hunk 1]
```diff
@@ -1230,66 +1230,134 @@ cargo check -Z unstable-options -Z check-cfg-well-known-values
 
 ### workspace-inheritance
 
-* RFC: [#2906](https://github.com/rust-lang/rfcs/blob/master/text/2906-cargo-workspace-deduplicate.md)
```
Comment (epage) [inline src/doc/src/reference/unstable.md:1233]: Please keep the RFC link like the other unstable features.

File: src/doc/src/reference/unstable.md [hunk 2]
```diff
@@ -1230,66 +1230,134 @@ cargo check -Z unstable-options -Z check-cfg-well-known-values
...
-* Tracking Issue: [#8415](https://github.com/rust-lang/cargo/issues/8415)
 
-The `workspace-inheritance` feature allows workspace members to inherit fields
-and dependencies from a workspace.
+### The `workspace.package` table
 
-Example 1:
+This would be in [`workspaces.md`][workspaces], under
```
Comment (epage) [inline src/doc/src/reference/unstable.md:1236]: Please prefix this with `Stablization:` and italicize it so its clear this is for the unstable-docs reader and not for what will be in the docs.

While the content makes it clear, being able to scan for information is important

File: src/doc/src/reference/unstable.md [hunk 3]
```diff
@@ -1230,66 +1230,134 @@ cargo check -Z unstable-options -Z check-cfg-well-known-values
...
+`[target."...".dependencies]` sections support the ability to reference the
+`[workspace.dependencies]` definition of dependencies.
+
+[inheriting-a-dependency-from-a-workspace]: #inheriting-a-dependency-from-a-workspace
+[workspaces]: workspaces.md
+[the-workspacemetadata-table]: workspaces.md#the-workspacemetadata-table
+
+### Inheriting a dependency from a workspace
+
+This would be in [`specifying-dependencies.md`][specifying-dependencies],
```
Comment (epage) [inline src/doc/src/reference/unstable.md:1318]: ditto about making this clear this is not for the final docs

File: src/doc/src/reference/unstable.md [hunk 4]
```diff
@@ -1230,66 +1230,134 @@ cargo check -Z unstable-options -Z check-cfg-well-known-values
...
+under [Renaming dependencies in Cargo.toml][renaming-dependencies-in-cargotoml]
```
Comment (epage) [inline src/doc/src/reference/unstable.md:1319]: Should this go earlier in the document where its talking about dependency sources?

In some ways, I see this filling that role but it also goes into a lot more detail than just sources and covers content from other sections in this document, making it fit at the end as well

File: src/doc/src/reference/unstable.md [hunk 5]
```diff
@@ -1230,66 +1230,134 @@ cargo check -Z unstable-options -Z check-cfg-well-known-values
...
-log2 = { version = "2.0.0", package = "log" }
-serde = { git = 'https://github.com/serde-rs/serde' }
-wasm-bindgen-cli = { path = "crates/cli" }
-```
+The `workspace.package` table is where you define keys that can be
+inherited by members of a workspace. These keys can be inherited by
+defining them in the member package with `{key}.workspace = true`.
+For Example:
 
 ```toml
```
Comment (epage) [inline src/doc/src/reference/unstable.md:1244]: Nit: I would not have a newline between `For example:` and the example but before `For example:`

File: src/doc/src/reference/unstable.md [hunk 6]
```diff
@@ -1230,66 +1230,134 @@ cargo check -Z unstable-options -Z check-cfg-well-known-values
...
-license-file.workspace = true
-keywords.workspace = true
-categories.workspace = true
-publish.workspace = true
 ```
 
+Keys that are supported:
+`authors`, `categories`, `description`, `documentation`, `edition`,
+`exclude`, `homepage`, `include`, `keywords`, `license`, `license-file`,
+`publish`, `readme`, `repository`, `rust-version`, and`version`.
```
Comment (epage) [inline src/doc/src/reference/unstable.md:1269]: Seems like this could be hard to parse as a comma separated list but unsure if any of the alternatives are better
- bulleted list
- table (where column/row is not significant)

File: src/doc/src/reference/unstable.md [hunk 7]
```diff
@@ -1230,66 +1230,134 @@ cargo check -Z unstable-options -Z check-cfg-well-known-values
...
+
+
+### The `workspace.dependencies` table
+
+The `workspace.dependencies` table is where you define dependencies to be
+inherited by members of a workspace. Inheriting dependencies was added as a
+way to reduce duplication of dependencies across members of a workspace.
```
Comment (epage) [inline src/doc/src/reference/unstable.md:1276]: I feel like the motivation isn't needed and is inconsistent since we didn't mention it in `workspace.package`

File: src/doc/src/reference/unstable.md [hunk 8]
```diff
@@ -1230,66 +1230,134 @@ cargo check -Z unstable-options -Z check-cfg-well-known-values
...
+Inheriting a dependency is as simple as setting the dependency in the workspace
+under `[workspace.dependencies]` as you would in`[dependencies]`. From there
+[specify that dependency][inheriting-a-dependency-from-a-workspace] in the member
+`[dependencies]` table.
```
Comment (epage) [inline src/doc/src/reference/unstable.md:1280]: Here we are saying its just like `[dependencies]` but then later we give restrictions on how to do it.  People might read this without going further to read the other.  We should be upfront that there are restrictions

File: src/doc/src/reference/unstable.md [hunk 9]
```diff
@@ -1230,66 +1230,134 @@ cargo check -Z unstable-options -Z check-cfg-well-known-values
...
+version = "0.2.0"
+
+[dependencies]
+log.workspace = true
+serde.workspace = true
+wasm-bindgen-cli.workspace = true
+```
+
+Dependencies in `[workspace.dependencies]` cannot be declared as `optional`.
+The `optional` key must be omitted or, if present, must be `false`. Dependencies
```
Comment (epage) [inline src/doc/src/reference/unstable.md:1307]: These two sentences feel redundant.  I know there is a slight distinction but we either should collapse them down or decide the distinction isn't worth it and leave off the second

File: src/doc/src/reference/unstable.md [hunk 10]
```diff
@@ -1230,66 +1230,134 @@ cargo check -Z unstable-options -Z check-cfg-well-known-values
...
+in the `[dependencies]`, `[dev-dependencies]`, `[build-dependencies]`, and
+`[target."...".dependencies]` sections support the ability to reference the
+`[workspace.dependencies]` definition of dependencies.
```
Comment (epage) [inline src/doc/src/reference/unstable.md:1310]: This feels weird being in the same paragraph as talking about `optional`.  

Some other random thoughts
- We should link to the specifying dependencies section
- What level of detail do we duplicate in here?  We say which tables support inheritance but leave all other details to the other document

File: src/doc/src/reference/unstable.md [hunk 11]
```diff
@@ -1230,66 +1230,134 @@ cargo check -Z unstable-options -Z check-cfg-well-known-values
...
+under [Renaming dependencies in Cargo.toml][renaming-dependencies-in-cargotoml]
+
+Dependencies can be inherited from a workspace. To do this, specify the
+dependency in the workspace's [`[workspace.dependencies]`][workspace.dependencies] table.
+After that add it to the `[dependencies]` table with `dep.workspace = true`.
+
+The `workspace` key cannot be defined with other keys that configure the source of the
+dependency. This means you cannot define `workspace` with keys like `version`, `registry`,
+`registry-index`, `path`, `git`, `branch`, `tag`, `rev`, or `package`. The `workspace`
+key can be combined with other keys, however:
```
Comment (epage) [inline src/doc/src/reference/unstable.md:1328]: 1. Also unsure about comma separated list here
2. This makes it sound like there are not other caveats (default-features)
3. We are also splitting up caveats, probably making it harder for users to find all the information they are looking for

File: src/doc/src/reference/unstable.md [hunk 12]
```diff
@@ -1230,66 +1230,134 @@ cargo check -Z unstable-options -Z check-cfg-well-known-values
...
+
+- `optional` - this introduces an optional dependency as usual, as well as a feature named
+  after the key (left hand side) of the dependency directive). Note that the
+  `[workspace.dependencies]` table is not allowed to specify `optional`.
```
Comment (epage) [inline src/doc/src/reference/unstable.md:1332]: Rahter than going into this much detail, should we just say it works like normal and link to the optional documentation?

File: src/doc/src/reference/unstable.md [hunk 13]
```diff
@@ -1230,66 +1230,134 @@ cargo check -Z unstable-options -Z check-cfg-well-known-values
...
+
+- `features` - this indicates, as usual, that extra features are being enabled over
+  the already-enabled features in the directive found in `[workspace.dependencies]`. The
+  result set of enabled features is the union of the features specified inline with
+  the features specified in the directive in the workspace table.
```
Comment (epage) [inline src/doc/src/reference/unstable.md:1337]: I feel like this could be more direct to help people more quickly see the pertinent information about this.

Maybe
```md
- `features`: These are additive with the features declared in the `[workspace.dependencies]`, see [...link to features...]
```

**Review by epage — COMMENTED** [review]
Thanks for getting this written up!

File: src/doc/src/reference/unstable.md (hunk 4 above)
Comment (Muscraft) [reply src/doc/src/reference/unstable.md:1319]: I thought about putting it in sources but decided against it. It also seems "more advanced" than just a source is why I put it where it is. I could go either way on this

File: src/doc/src/reference/unstable.md (hunk 6 above)
Comment (Muscraft) [reply src/doc/src/reference/unstable.md:1269]: I think a table may be better with just 2 columns and make it alphabetical like it is now

File: src/doc/src/reference/unstable.md (hunk 7 above)
Comment (Muscraft) [reply src/doc/src/reference/unstable.md:1276]: That works for me

File: src/doc/src/reference/unstable.md (hunk 8 above)
Comment (Muscraft) [reply src/doc/src/reference/unstable.md:1280]: How do you think it should be worded?

File: src/doc/src/reference/unstable.md (hunk 9 above)
Comment (Muscraft) [reply src/doc/src/reference/unstable.md:1307]: Dropping the second one seems better. It is the more restrictive of the two

File: src/doc/src/reference/unstable.md (hunk 10 above)
Comment (Muscraft) [reply src/doc/src/reference/unstable.md:1310]: I think linking might be better here

File: src/doc/src/reference/unstable.md (hunk 11 above)
Comment (Muscraft) [reply src/doc/src/reference/unstable.md:1328]: We can make it a table and move `default-features` to it. It might lose a little bit of explanation but it should be acceptable

File: src/doc/src/reference/unstable.md (hunk 13 above)
Comment (Muscraft) [reply src/doc/src/reference/unstable.md:1337]: That is much less verbose but gets the point across well, I like it

File: src/doc/src/reference/unstable.md (hunk 8 above)
Comment (epage) [reply src/doc/src/reference/unstable.md:1280]: I actually suspect the whole thing can go.  The prior sentence (The `workspace.dependencies` table is where you define dependencies to be
inherited by members of a workspace. ) coneys most of the information.  I'd then have the behavior (what tables can inherit, what can't be specified, that features is merged) in the following paragraph and then examples

File: src/doc/src/reference/unstable.md [hunk 14]
```diff
@@ -1233,63 +1233,147 @@ cargo check -Z unstable-options -Z check-cfg-well-known-values
...
+| A - I           | J - Z          |
+|-----------------|----------------|
+| `authors`       | `keywords`     |
+| `categories`    | `license`      |
+| `description`   | `license-file` |
+| `documentation` | `publish`      |
+| `edition`       | `readme`       |
+| `exclude`       | `repository`   |
+| `homepage`      | `rust-version` |
+| `include`       | `version`      |
```
Comment (epage) [inline src/doc/src/reference/unstable.md:1256]: I think most people would expect a table like this to read left-to-right.  The columns would not have significance

File: src/doc/src/reference/unstable.md [hunk 15]
```diff
@@ -1233,63 +1233,147 @@ cargo check -Z unstable-options -Z check-cfg-well-known-values
...
+
+ The `workspace` key cannot be defined with:
+
+| A - P              | Q - Z            | 
+|--------------------|------------------|
+| `branch`           | `registry`       |
+| `default-features` | `registry-index` |
+| `git`              | `rev`            |
+| `package`          | `tag`            |
+| `path`             | `version`        |
```
Comment (epage) [inline src/doc/src/reference/unstable.md:1346]: ditto about table formatting

File: src/doc/src/reference/unstable.md (hunk 14 above)
Comment (epage) [inline src/doc/src/reference/unstable.md:1256]: Should we call out the path behavior?
- license-file and readme are relative to the workspace root
- `include` and `exclude` are relative to your package root

File: src/doc/src/reference/unstable.md [hunk 16]
```diff
@@ -1233,63 +1233,147 @@ cargo check -Z unstable-options -Z check-cfg-well-known-values
...
+The `workspace.dependencies` table is where you define dependencies to be
+inherited by members of a workspace. Inheriting a dependency is as simple as
+[specifying the dependency][specifying-dependencies] in this table then
+[inheriting that dependency][inheriting-a-dependency-from-a-workspace] in the member
+`[dependencies]` table.
+
+Dependencies from this table can be referenced by `[dependencies]`, 
+`[dev-dependencies]`, `[build-dependencies]`, and `[target."...".dependencies]`.
+Dependencies from this table cannot be declared as `optional`.`features` declared
+in this table are additive with the `features` from `[dependencies]`, see [features]
```
Comment (epage) [inline src/doc/src/reference/unstable.md:1293]: Reading over this, I feel like we are burying the lede.  Do we need the first sentence and could we fit the other sentences in as bullet points?

File: src/doc/src/reference/unstable.md [hunk 17]
```diff
@@ -1233,63 +1233,147 @@ cargo check -Z unstable-options -Z check-cfg-well-known-values
...
 ```
 
+
+### The `workspace.dependencies` table
+
+The `workspace.dependencies` table is where you define dependencies to be
+inherited by members of a workspace. Inheriting a dependency is as simple as
+[specifying the dependency][specifying-dependencies] in this table then
+[inheriting that dependency][inheriting-a-dependency-from-a-workspace] in the member
+`[dependencies]` table.
```
Comment (epage) [inline src/doc/src/reference/unstable.md:1288]: This second sentence still needs some re-work.  "its as simple as" implies its the same but its not.  I'd at least say "similar to"  and make this the introduction to the caveats below

File: src/doc/src/reference/unstable.md [hunk 18]
```diff
@@ -1233,63 +1233,147 @@ cargo check -Z unstable-options -Z check-cfg-well-known-values
...
+[workspaces]: workspaces.md
+
+
+### Inheriting a dependency from a workspace
+
+*Stabilization*: This would be in [`specifying-dependencies.md`][specifying-dependencies],
+under [Renaming dependencies in Cargo.toml][renaming-dependencies-in-cargotoml]
+
+Dependencies can be inherited from a workspace. To do this, specify the
+dependency in the workspace's [`[workspace.dependencies]`][workspace.dependencies] table.
```
Comment (epage) [inline src/doc/src/reference/unstable.md:1334]: I feel like this sentences can be merged.  Instead of telling someone that they can do something and then how, just tell them how, that implies they can.

File: src/doc/src/reference/unstable.md [hunk 19]
```diff
@@ -1233,63 +1233,147 @@ cargo check -Z unstable-options -Z check-cfg-well-known-values
...
+|--------------------|------------------|
+| `branch`           | `registry`       |
+| `default-features` | `registry-index` |
+| `git`              | `rev`            |
+| `package`          | `tag`            |
+| `path`             | `version`        |
+
+The `workspace` key can be defined with:
+- [`optional`][optional] -  Note that the`[workspace.dependencies]` table is not allowed to specify `optional`.
+- [`features`][features] - These are additive with the features declared in the `[workspace.dependencies]`
```
Comment (epage) [inline src/doc/src/reference/unstable.md:1350]: Unless there is a cargo style thing I've not noticed, then I'd recommend `:` rather than `-` for separating out an item from the content for it.

**Review by epage — COMMENTED** [review]
Documentation is always the hardest part but we're getting closer!

File: src/doc/src/reference/unstable.md (hunk 14 above)
Comment (Muscraft) [reply src/doc/src/reference/unstable.md:1256]: I'll change it to this

Comment (Muscraft) [reply src/doc/src/reference/unstable.md:1256]: I will add it, it seems like a good thing to have

File: src/doc/src/reference/unstable.md (hunk 16 above)
Comment (Muscraft) [reply src/doc/src/reference/unstable.md:1293]: I think thats possible

File: src/doc/src/reference/unstable.md [hunk 20]
```diff
@@ -1233,63 +1233,149 @@ cargo check -Z unstable-options -Z check-cfg-well-known-values
...
+
+The `workspace.dependencies` table is where you define dependencies to be
+inherited by members of a workspace. 
+
+To inherit a dependency:
+- [specify the dependency][specifying-dependencies] in this table
+  - Dependencies from this table cannot be declared as `optional`
+  - [`features`][features] declared in this table are additive with the `features` from `[dependencies]`
+- [inherit that dependency][inheriting-a-dependency-from-a-workspace] in the member
+`[dependencies]` table.
```
Comment (epage) [inline src/doc/src/reference/unstable.md:1295]: I feel like the outer bullet list loses the association of the two steps

What about
```
> Specifying a workspace dependency is similar to [package dependencies][...] except:
> - ...
> - ...
> You can then [inherit the workspace dependency as a package dependency][...]
```

Comment (Muscraft) [reply src/doc/src/reference/unstable.md:1295]: That is much better!
